### PR TITLE
add PLATFORM env var support to makefile

### PR DESCRIPTION
**Pull request by juanvallejo**
Allow a `PLATFORM` env var to be set in order to allow building for different platforms.

Usage: 
```
$ PLATFORM=darwin make apiserver
docker run --rm -ti -v /home/jvallejo/go/src/github.com/kubernetes-incubator/service-catalog:/go/src/github.com/kubernetes-incubator/service-catalog -v /home/jvallejo/go/src/github.com/kubernetes-incubator/service-catalog/.pkg:/go/pkg scbuildimage env GOOS=darwin GOARCH=amd64 go build -i  -ldflags "-X github.com/kubernetes-incubator/service-catalog/pkg.VERSION=e40e7c0-dirty" -o bin/apiserver github.com/kubernetes-incubator/service-catalog/cmd/apiserver
```

TODO: update docs to mention this

cc @pwittrock

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -42,7 +42,16 @@ NEWEST_GO_FILE = $(shell find $(SRC_DIRS) -name \*.go -exec $(STAT) {} \; \
                    | sort -r | head -n 1 | sed "s/.* //")
 TYPES_FILES    = $(shell find pkg/apis -name types.go)
 GO_VERSION     = 1.7.3
-GO_BUILD       = env GOOS=linux GOARCH=amd64 go build -i $(GOFLAGS) \
+
+ifndef PLATFORM
+	PLATFORM = linux
+endif
```
Comment (duglin) [inline Makefile]: I think you can just do:
```
PLATFORM?=linux
ARCH?=amd64
```

**Comment by juanvallejo** [comment]
@duglin thanks for the feedback, that looks much cleaner

**Comment by krancour** [comment]
I am strongly opposed to this. We actually used to support this and we eliminated it for good reason. The rationale went something like this: We have no CLI component or anything that a user might _reasonably_ need to run from various and sundry OS/architectures. All we have are components that are intended to run _inside_ a Kubernetes cluster. Necessarily, anything running _inside_ a k8s clusters is containerized, and consequently, we can safely assume uses linux/amd64. (This discounts the extremely outlying scenario where people are running clusters on linux/arm.)

We decided that, given the above, we'd shed _official_ support for builds for different OS/architectures since it was only going to encourage running the components out-of-cluster, which is a configuration we haven't really anticipated, tested with, or care to support.

Also note that the implementation offered in this PR isn't sufficient. If we _were_ going to compile binaries for multiple OS/architecture, they belong in their own directories, e.g. `bin/linux/amd64` to ensure they're distinguishable from one another. If that _isn't_ done, someone who has cross-compiled for a non-linux/amd64 OS/arch could have a darwin/64 build (for instance) hanging out in `bin/` when the Docker image is produced. The Docker build cannot distinguish that `bin/controller` (for instance) isn't linux/amd64 and you end up building an image that won't ever work. To guard against this, as I said, you need to introduce some additional directory structure and then, much worse, you need to start getting a little fancier with build target dependencies so that when doing the Docker build, the build system verifies that the _right_ binaries have been built. This is the complexity I said we had wished to shed.

**Comment by juanvallejo** [comment]
@krancour 
> We actually used to support this and we eliminated it for good reason. The rationale went something like this: We have no CLI component or anything that a user might reasonably need to run from various and sundry OS/architectures. All we have are components that are intended to run inside a Kubernetes cluster. 

This makes sense, and if cross compiling support was already removed in the past, then I agree it would make no sense to reintroduce it.

Closing, thanks for the feedback.

**Comment by duglin** [comment]
FYI - we do care about other platforms but my assumption was that we would do those builds on those platforms and wouldn't need to leverage golang's cross complication feature.

**Comment by juanvallejo** [comment]
> FYI - we do care about other platforms but my assumption was that we would do those builds on those platforms and wouldn't need to leverage golang's cross complication feature.

Right, but what this PR aimed to at least do was to avoid changing the value of `GOOS` manually [in the Makefile](https://github.com/kubernetes-incubator/service-catalog/blob/master/Makefile#L45) when compiling in mac, for example. Otherwise, you'd just end up with a linux binary every time, even if compiling on mac. Maybe a simpler solution would just be to check if `GOOS` is already defined in a user's environment, and use that instead of hard-coding `linux`.

**Comment by duglin** [comment]
I (think) I'd be in favor of dropping the GOOS and GOARCH from the "go build" command - then just let go get it from the platform its running on.

**Comment by krancour** [comment]
> I (think) I'd be in favor of dropping the GOOS and GOARCH from the "go build" command - then just let go get it from the platform its running on.

I'll re-iterate that it doesn't make a ton of sense to build any of our components for non-linux/amd64 in the first place, because they're all intended to run in-cluster, which means in-container, which means linux/amd64 (or in _very rare cases we don't care about_, linux/arm).

Additionally, dropping `GOOS` and `GOARCH` from the `go build` command so that compilation occurs for the native architecture will introduce new problems into our build. Specifically, someone on a Mac who does `NO_DOCKER=1 make build images`, will end up with Docker images that have Darwin binaries on them that won't execute.

**Comment by duglin** [comment]
@krancour IBM will want to build/run everything on Power and Z systems.

**Comment by krancour** [comment]
> IBM will want to build/run everything on Power and Z systems.

Interesting. That's definitely a use case I hadn't foreseen-- forgot Linux on z/OS was even a thing. It has Docker too? And there are z/OS versions of all k8s binaries, Docker images for addons (like DNS), etc.?

At any rate, this is enough to make me re-consider the value of this PR.

That being said, I think there's tremendous potential for people to shoot themselves in the foot with this. By way of example, if this PR is merged, the following command will succeed, but will produce non-viable Docker images.

```
$ PLATFORM=darwin make build images
```

Knowing individual mileage may vary so significantly, I'd suggest we deliberately do not document this so as not to encourage it. @duglin if that sounds reasonable to you, then I will offer my LGTM.

**Comment by duglin** [comment]
> forgot Linux on z/OS was even a thing

ouch!  :-)

Yes, we've ported Docker to P and Z.

As I said earlier, I'm not sure we should allow for people to pass-in options to do cross compiles - at least not yet because I don't fully grok the implications.  So I think a good first step is to just remove the stuff that forces the build to x86 so that I can take the existing stuff and try, w/o modifications, building on P&Z.

**Comment by krancour** [comment]
Ok... but that suggestion descends back into more controversial territory...

If we stop passing GOOS and GOARCH to the build, building out-of-container using `NO_DOCKER=1` (something Google wanted-- which I'll admit I occasionally use for speed) will build for _native_ os/arch (which is what you want), but on _many_ systems (e.g. Macs), that will produce binaries that can't be bundled up into the Docker images.

Some frustration is going to show through here and I apologize for that. Honestly, I think we're entertaining too many degrees of freedom with this build. The major benefit of containerizing builds was that with a minimal toolchain installed on the native system, all contributors could build/test in a uniform and "hermetically sealed" development environment. All these degrees of freedom are negating those benefits and making the build more complex and more brittle. My impulse is to eliminate permutations instead of introducing new ones.

**Comment by duglin** [comment]
Not being able to use service catalog on or P/Z build of kube is a non-starter for us.
Let's discuss this on the next weekly call.

**Comment by arschles** [comment]
@duglin I don't think anyone is against trying to add the ability to build service-catalog on systems that people need to build on. Our big concern is that our build & release system does not get more complex for developers to use and improve later.

As I write this, we have 9 full business days until our MVP-1 is due, so I want everyone to be more concerned with getting service-catalog working than anything else. With that in mind, continuing the discussion here is not going to move us closer to a MVP quality product. Please submit a PR for the P/Z support so we can continue this discussion there, in the context of a concrete implementation. I'm confident we can all agree on something that addresses the complexity concerns above.

**Comment by pmorie** [comment]
> Not being able to use service catalog on or P/Z build of kube is a non-starter for us.

Does this PR remove that ability?

**Comment by krancour** [comment]
I am ok with this PR as long as we don't document / advertise / encourage people to do this, because as I've stated, unless you really know what you're doing, individual mileage will vary.

I think @duglin prefers a different approach than this PR (because of concerns over the viability of cross-compiling for P/Z?) He'd prefer to _always_ build for the native architecture. For most contributors, that means a build in-container will yield linux/amd64, but a build _out-of-container_ (which some folks do, even though I caution against that as well) could yield binaries (e.g. for Mac) that _won't_ run when the `images` target bundles them onto a Docker image.

Bottom line... I think we should re-open, review, and merge this PR and worry _later_ about whether cross-compiling for P/Z is really enough of an issue that we want to consider modifying the build in a way that will give people substantially more rope with which to hang themselves.

Executive decision... I am re-opening and granting this my LGTM. Others can argue it further, but I've had enough debate, have said my peace and _I'm_ cool with this.

**Comment by duglin** [comment]
Just one comment:
> that means a build in-container will yield linux/amd64

this may be true today w/o any changes but at some point these base images (like `golang`) will hopefully be multi-arch enabled. Meaning the docker daemon will download the arch-specific version of the image automatically. Which means the build on a Z system will yield a Z docker image w/o the need for a customized Dockerfile (at least for the FROM part of it).

**Comment by krancour** [comment]
I think that's beside the point though. I'm more worried about what can happen in the _out-of-container_ build if we went with your alternative proposal of always building for the native os/arch.

In my own case, I'd consistently end up with Darwin binaries that `make images` could successfully bundle into images... but then those images would never work.

You can paint yourself into a similar corned with the option exposed by this PR, but you'd at least be doing so with more awareness since you would have explicitly selected a non-default OS/arch.

**Comment by duglin** [comment]
We can live with this one.
LGTM

**Comment by krancour** [comment]
I've sensed lately that @pmorie and @arschles are pressing for less churn (and less bike shedding) on the build process in favor of more work and discussion on features... so rather than expose this to further scrutiny, I'm acting in good faith and making the decision to merge this with just the two LGTMs it currently has.

**Comment by pmorie** [comment]
👍
